pkg/asset/installconfig/aws/validation.go

```diff
@@ -166,6 +184,167 @@ func validateAMI(ctx context.Context, config *types.InstallConfig) field.ErrorLi
+	}
+	return nil
+}
+
+func validateEIPAllocations(client ec2iface.EC2API, ctx context.Context, meta *Metadata, fldPath *field.Path, config *types.InstallConfig) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if config.Platform.AWS.EIPAllocations != nil && len(config.Platform.AWS.EIPAllocations.IngressNetworkLoadBalancer) == 0 {
+		return nil
+	}
+	var eipAllocations []awstypes.EIPAllocation
+	if config.Platform.AWS != nil && config.Platform.AWS.EIPAllocations != nil && len(config.Platform.AWS.EIPAllocations.IngressNetworkLoadBalancer) == 0 {
+		eipAllocations = config.Platform.AWS.EIPAllocations.IngressNetworkLoadBalancer
+	}
+	var allocationIDs []string
+	for _, allocationID := range eipAllocations {
+		allocationIDs = append(allocationIDs, string(allocationID))
+	}
+
+	// Validate EIP allocation IDs
+	if err := validateEIPAllocationsList(allocationIDs); err != nil {
+		fmt.Printf("Validation error: %v\n", err)
+	} else {
+		fmt.Println("All allocation IDs are valid.")
+	}
+
+	// Check if eipAllocations exist
+	for _, allocationID := range allocationIDs {
+		_, err := eipAllocationExists(client, allocationID)
+		if err != nil {
+			err = fmt.Errorf("eipAllocations does not exists due to error %s:", err)
+			return append(allErrs, field.InternalError(fldPath, err))
+		}
+	}
+
+	// Get all associated EIPs in the list
+	associatedEIPs, err := getAssociatedEIPs(client, allocationIDs)
+	if err != nil {
+		err = fmt.Errorf("could not get check if eips are associated due to error %s:", err)
+		return append(allErrs, field.InternalError(fldPath, err))
+	}
+	fmt.Printf("Asociated eips %v", associatedEIPs)
+
+	// Print error if EIPs are associated
+	if len(associatedEIPs) > 0 && config.Platform.AWS != nil && config.Platform.AWS.EIPAllocations != nil && len(config.Platform.AWS.EIPAllocations.IngressNetworkLoadBalancer) == 0 {
+		return append(allErrs, field.Invalid(fldPath, associatedEIPs, fmt.Errorf("eipAllocations %s can't be used as %s are already associated", config.Platform.AWS.EIPAllocations.IngressNetworkLoadBalancer, associatedEIPs).Error()))
+	}
+
+	// BYO Subnets
+	if len(config.Platform.AWS.Subnets) != 0 {
+		if err = compareBYOSubnetsWithEIPAllocations(config); err != nil {
+			return append(allErrs, field.Invalid(fldPath, nil, compareBYOSubnetsWithEIPAllocations(config).Error()))
+		}
+	} else { // Installer Created Subnets
+		allzones, err := meta.AvailabilityZones(ctx)
+		if err != nil {
+			return append(allErrs, field.InternalError(fldPath, err))
+		}
+		if err = compareInstallerCreatedSubnetsWithEIPAllocations(config, allzones); err != nil {
+			return append(allErrs, field.Invalid(fldPath, nil, compareInstallerCreatedSubnetsWithEIPAllocations(config, allzones).Error()))
+		}
+	}
+	return nil
+}
+
+func compareBYOSubnetsWithEIPAllocations(config *types.InstallConfig) error {
+	if len(config.Platform.AWS.Subnets) != len(config.Platform.AWS.EIPAllocations.IngressNetworkLoadBalancer) {
```

Comment:
Though this works for 95% of users, we should be aware that this an over simplification of validation as it assumes a "clean" BYOB VPC, and something we haven't quite came to an agreement on in the EP.

When there are other non-tagged subnets in a VPC that the AWS CCM will generously select, there are two possible outcomes:
1. *Unintentional*: The user sets `len(subnets) == len(EIPs)` correctly, validation passes, but the cluster fails to install due to IngressController LB failing to provision (the unintentionally AWS CCM selected > subnets than EIPs).
2. *Intentional*: The user *intentionally* configures their non-cluster VPC subnet tags so that the AWS CCM selects more subnets than `len(installconfig.aws.subnets)` for some edge case (hybrid cloud?), and they can't pass validation with EIPs.
 
I think # 1 is more likely than # 2 to happen. To be clear, I'm not saying we have to be perfect here (that's the installer team's call), but I think it's pretty important to understand the implications. The solution is to count all untagged (or tagged with the `cluster-id`) subnets, filter out public/private, and compare that subnet count here (like we discussed in the EP). I probably over complicated it in my earlier suggestions. We could have a common function that provides the `getActualDefaultIngressControllerSubnets`.

But if we are okay to keep this type of validation for now, we should consider:
1. Having the installer warn that "things might not work as expected" when it finds a non-tagged (or already tagged with the `cluster-id`) subnet
2. Changing to `len(config.Platform.AWS.Subnets) <= len(config.Platform.AWS.EIPAllocations.IngressNetworkLoadBalancer)` to enable potential edge case # 2 (because AWS will only *add* more subnets).
   - This avoids falsely rejecting valid configurations, but at the expense of the "clean" BYOB VPC (no untagged subnets) that might accidentally provide `len(EIPS) > subnets` and the cluster install fails. This feels more "correct" to me.

@patrickdillon @mtulio do you have thoughts?

Reply:
Not the AWS expert here. 
While performing step 1(of the proposed solution/validation), we should be able to detect the "clean" BYOB VPC (no untagged subnets) since it appears we would be iterating through the subnets. If yes, can we use that information to allow `len(EIPS) > subnets` only when "clean" BYOB VPC is detected?

Reply:
Hi @gcs278, I like the idea of "alerting" the user for possible failed configurations, although I am not sure if we will extrapolate the responsibility of installer here when trying to discover more subnets when user choose to "read/touch only in those subnets I've added to the config", specially considering this is a problem on CCM that became the "normal" behavior.

Perhaps adding a "red" admonition in the ingress EIP documentation page when using this feature rather than add this validation?

Reply:
@gcs278 what if someone provides less eips than the number of subnets ?

Reply:
> what if someone provides less eips than the number of subnets ?

I think the cluster will finish installing, but it will report failed because the Ingress CO failed to go Available. The `default` IngressController's load balancer will fail to provision. Though it's not optimal, a cluster admin could fix the issue in Day 2, so it's not a catastrophic fatal. You should double check that though.

There's another problem I overlooked with validation here . When installing a public cluster, you typically provide  public *and* private subnets. You can also specify localzone, wavelength, outpost zone subnets which AWS CCM [filters out](https://github.com/openshift/cloud-provider-aws/blob/f02a2a760895cc33753bfc130f3d08826563ebfa/pkg/providers/v1/aws.go#L1727-L1732). I think you could also provide multiple subnets in the same AZ which would break the logic. So, even in a "clean" VPC with fully-tagged subnets, the current validation still won't work correctly.

You could argue that we are effectively trying to duplicate the [validation](https://github.com/openshift/cloud-provider-aws/blob/1ac3e5418ebef3f408dd3f34781a1f3da2ed0895/pkg/providers/v1/aws_loadbalancer.go#L173-L175) that the AWS CCM does. The problem though is that we need to mostly repeat the AWS CCM's subnet discovery logic (public vs. private, subnet-type filtering, one-per-AZ...) to count the right number of subnets.

I think I've come around to suggesting that we punting on validation for now, given that:
1. Even when provided BYO Cluster Subnets, it's still fragile and complex to count the number of subnets that subnet discovery will pick for the Load Balancer (it's possible, but the ROI isn't there for the effort)
2. The AWS CCM already does this validation in Day 2, we'd be trying to repeat it in Day 0.
3. The cluster still finishes installing, giving cluster admins an opportunity to fix it day 2 (sub-optimal, but not catastrophic)

Thoughts?